### PR TITLE
Improvements to GitHub Actions CI: integration tests on PR approval, code coverage

### PR DESCRIPTION
**Pull request by lbianchi-lbl**
# Improvements to GitHub Actions CI: integration tests on PR approval, code coverage

## Summary

- Run integration tests only when a PR is approved by reviewers
  - Similarly to what we previously had in CircleCI, integration tests are run when the `approved` label is set on a PR
  - The novel part is that the `approved` label is set or removed automatically from the total count of approving reviews any time a review for that PR is submitted or modified
    **UPDATE 2020-12-03** This functionality will be incorporated separately in its own PR #111
- Code coverage reports are generated and submitted to Codecov, which provides interesting statistics and coverage checks (testing positive increase in coverage) that can be required to pass as a condition for mergeability

## Changes proposed in this PR

- Extract common tasks to separate actions (under this repository's `.github/actions` directory)
- Split CI tests in 3 separate workflows
  - `pr.yml`, with rapid checks, to be run whenever:
    - A PR is created (or undrafted)
    - Changes (commits) are pushed to a PR branch
  - `pr-approved.yml`, including integration tests, to be run whenever a PR is labeled as approved
  - **UPDATE 2020-12-03** See PR #111 <del>`pr-review.yml`: automatically synchronizes PR approval status (by adding or removing approved label) whenever a review for that PR is submitted or updated</del>
  - `main.yml`, running the most comprehensive test suite on protected branches, on schedule or on manual trigger
- Reinstate support for code coverage report generation, analysis and checks via Codecov

## Status

This setup has now been tested on my fork, and, assuming nothing is obviously broken, is ready for review. **UPDATE 2020-12-03** The automatic management of the `approved` label will be functional only after merging #111. In the meantime, this can already be merged in: all workflows will work just fine, with the only difference being that the `approved` label has to be applied manually.

Let me know if you have any feedback or questions. I've added some details below about the motivation and choices behind the solution adopted in this PR. As usual, any aspect of this is open for comments or discussions.

### To review/discuss

- [x] Should Python 3.8 be added to the test matrix? From @jsiirola's original review comment:
    > Should we add 3.8? (given that 3.9 is out now and 3.8 is the "standard" anaconda distribution)

### Post-merge tasks

- [ ] Update required checks in Branch Protection Rules (via admin panel)
- [ ] Verify that the `main.yml` workflow runs on the default (`main`) branch both on schedule and on demand
- [ ] Merge in changes from `main` into existing PRs
- [ ] Clean up previous workflow runs in Actions panel
- [ ] Verify that the badge in the README works
- [ ] Verify/showcase Codecov functionality (PR summary and checks) using minimal PR (TODO)
- [ ] Add Codecov checks (positive coverage delta at the PR patch or whole project level) as required for merging PR branch
- [ ] Add Codecov badge to README
  - We should be able to do this without conflict without needing to open another PR (assuming existing PRs do not have changes to the README)

---

### Runtime of various workflow jobs

(in minutes)

- Tests (Linux and Windows, Windows is slower)
  - No integration: 10-15
  - Integration: 30-45
  - Complete: 40-60
- Build Sphinx docs: < 5 minutes
- pylint: < 5 minutes
- Tests with coverage (Linux only)
  - No integration: 18-20
  - Integration: 40-45
  - Complete: 60-70

#### Goal

Develop an optimal workflow strategy for running integration tests that:

- Ensures that all required checks are run to determine mergeability
- Minimizes excessive runtime and manual caretaking

#### Requirements and constraints

- Test workflow for PR:
  - Run non-integration tests and other checks (~10-12 mins) at every PR update
  - Run integration tests as a final check before the PR can be merged
- All checks (integration and non-integration tests, docs build, static analysis) should pass before the PR can be merged
  - Use the GitHub required checks functionality on protected branches to enforce mergeability
    - The chosen workflow strategy should support this
- GitHub Actions (GHA) limitations/idiosyncrasies
  - A workflow run will be initiated every time the trigger condition is matched
    - Conditional job execution (skipping) can provide a way to make workflow trigger criteria more restrictive
    - Nonetheless, if a too frequent "primary trigger" is chosen, a large number of partial workflow runs will be created, resulting in e.g. spamming the workflow run dashboard
  - To be picked up correctly as branch protection checks, workflows must be triggered by `pull_request` events
    - => Workflow can be triggered by other events (e.g. `pull_request_reviews`) by generating an intermediate "bridge" `pull_request` trigger event
  - Successive runs of the same job for a given check will reset or invalidate existing runs for that check
    - => Integration tests and non-integration tests cannot be defined in the same workflow with conditional skipping
  - Unlike e.g. CircleCI, job-level configuration reuse is not supported
    - => Custom actions can be used to reuse repeated sequences of steps in a DRY fashion
  - There is no native way in either GHA or the GitHub API to determine a PR's overall approval status (i.e. if the PR has the required number of approval)
    - => A PR's approval status can still be computed programmatically by counting the number of approved reviews among the most recent reviews

#### Solution adopted in this PR

- Use separate workflows with simple trigger conditions rather than a single workflow with complex conditional job execution logic
  - The disadvantage of having to redefine slight variations of the same job in each workflow file is mitigated by using actions to group most of the repeated tasks
- Use a `labeled` (i.e. "label added") event with the special `approved` label to trigger integration tests
  - Being a `pull_request` event, it matches the requirement for branch protection checks
  - Easy to generate programmatically and therefore suited to be used as a "bridge" event to be triggered from other workflows
  - Can be also triggered manually in a simple way (by removing and re-adding the `approved` label directly from the web interface) if need be
  - The workflow is triggered whenever a PR is labeled; however, it's skipped unless the added label is `approved`
  - **Alternatives** In principle, other sufficiently infrequent `pull_request` event types can be used as the "bridge" event, e.g. assigning or requesting a review from a bot account such as @idaes-build
- Use the JavaScript bindings to the GitHub API provided by actions/github-script to check and update the PR approval status by applying the `approved` label
  **UPDATE 2020-12-03**: See PR #111
  <del>
  - This had to be implemented as ready-made solutions were not suitable and resulted in mismatch with the PR approval status as computed by GitHub (i.e. the one displayed on the web interface)
    - The main limitation is that a naive criterion of counting how many reviews have an `APPROVED` state does not take into account that a user's review can be superseded by newer reviews with a different approval status
  - The official GitHub API bindings make it easy to both query and update the PR's metadata programmatically
  - The inline JavaScript script in `pr-review.yml` is arguably a reasonable compromise between a less readable shell equivalent and a full-fledged custom Node.js action (which, among other things, would require bundling all its Node.js dependencies in the repo)
  </del>
- Use a separate `main.yml` workflow to run tests on the protected branches
  - This runs the complete test suite (plus other checks, e.g. building the docs)
  - The result of this workflow is displayed as the main repository build/"checks passing" status in the README
  - Currently run on:
    - Schedule (once a day, nightly style) on the `main` branch
    - Manual trigger (from the Actions dashboard or via an API call) on the `main` branch
    - Commits pushed to `rel_*` branches
  - Currently not run when commits are pushed on the `main` branch (e.g. when merging in a PR)
    - This is up to discussion and can easily be activated (by adding `main` to the list of branches patterns in `on.push.branches`)
    - The rationale for this choice being:
      - This could potentially result in "CI congestion" when merging multiple PRs within a short time
      - The `main` branch, being the default branch, is covered by the `main.yml` workflow running on schedule or triggered manually (currently: once a day), which can also be triggered on demand when needed

#### Description of a PR workflow

(D: developer, R1, R2: reviewers)

This can also serve as a reference for the expected behavior, since this would otherwise be tricky to test programmatically.

**UPDATE 2020-12-03** The general idea of this workflow should be still valid, but the implementation details will likely change. The PR review approval workflow is described in detail in PR #111

- D creates a PR
  - `pr.yml` triggers
    - The rapid (non-integration) checks are run
  - The PR cannot be merged (no reviews, no integration tests)
- D pushes new commits
  - `pr.yml` triggers
    - The rapid checks are run
  - The PR still cannot be merged
- D requests reviews from R1 and R2
  - No further tests are run (since the code hasn't changed)
- R1 leaves an "approved" review
  - `pr-review.yml` triggers
    - The PR has 1 approved reviews, so no further action is taken
  - The PR still cannot be merged
- R2 leaves a "changes requested" review
  - `pr-review.yml` triggers
    - The PR has 1 approved review, so no further action is taken
  - The PR still cannot be merged
- D implements the requested changes and pushes them to the PR branch
  - `pr.yml` triggers
    - The rapid checks are run
  - R1 and R2's reviews are dismissed
    - `pr-review.yml` triggers
      - The PR has now 0 approved reviews, so no further action is taken
- R1 leaves an "approved" review
  - `pr-review.yml` triggers
    - The PR has now 1 approved reviews, so no further action is taken
- R2 leaves an "approved" review
  - `pr-review.yml` triggers
    - The PR has now 2 approved reviews
    - The `approved` label is not present, so it is added
  - The PR still cannot be merged because the integration tests haven't been passed
  - `pr-approved.yml` triggers on adding the `approved` label
    - Integration tests are started
  - The PR still cannot be merged because the integration tests are still running
- If the integration tests pass: the PR can now be merged
- If the integration tests fail:
  - D pushes changes to the PR branch
    - `pr.yml` triggers, rapid checks are run
  - Existing reviews are dismissed
    - `pr-review.yml` triggers
      - The PR has now 0 approved reviews
      - The `approved` label is present when it shouldn't, so it is removed
    - The PR review process starts again
      - Integration tests are not run until the PR is approved again

### CODE REVIEW

File: .github/workflows/pr-approved.yml
Comment (jghouse88) [inline .github/workflows/pr-approved.yml:13]: @lbianchi-lbl Quick question: do we  need to add the approved label manually or is this automatically added once two approving reviews are available?

Comment (lbianchi-lbl) [reply .github/workflows/pr-approved.yml:13]: Yes, the `approved` label is managed automatically by the `pr-review.yml` workflow:

- Added when the second approving review is submitted
- Removed if the label is present and there are fewer than 2 approving reviews
- Left as-is in the remaining cases

File: .github/workflows/main.yml
Comment (lbianchi-lbl) [reply .github/workflows/main.yml:30]: Good point. Personally I think it'd be good to include 3.8. Let's wait until tomorrow to see if there are other thoughts on this, otherwise I'll add it.

File: .github/workflows/main.yml
Comment (lbianchi-lbl) [reply .github/workflows/main.yml:100]: Good catch, thanks. For some reason I understood that the tokes was needed for GitHub Actions regardless of the repo's access status, but you're right, the Codecov docs explicitly mention GHA among the CI providers that do not require a token for public repos. Getting rid of that now.

**Review by jsiirola — APPROVED** [review]

**Review by jsiirola — DISMISSED** [review]

**Comment by jghouse88** [comment]
@lbianchi-lbl do you need a rubber stamp approval to merge this?

**Comment by lbianchi-lbl** [comment]
@jghouse88 thanks for asking. Yeah, I think at this point this is ready to go. I might still wait until tomorrow before doing the actual merge, since I'd like to have some dedicated time to go through the post-merge tasks and make sure that everything works smoothly.

**Review by jghouse88 — APPROVED** [review]
Rubber stamp LGTM

**Comment by jghouse88** [comment]
@lbianchi-lbl Done.

**Comment by dangunter** [comment]
+1 for adding 3.8 to the list; looking forward to dropping 3.6!

File: .github/workflows/main.yml
Comment (dangunter) [reply .github/workflows/main.yml:30]: I agree, let's add 3.8!

**Review by dangunter — CHANGES_REQUESTED** [review]

File: .github/workflows/main.yml
Comment (lbianchi-lbl) [reply .github/workflows/main.yml:34]: :heavy_check_mark:

**Comment by lbianchi-lbl** [comment]
> I think this is fine, but I do have a question about coverage: why do you not always collect coverage for all pytest runs instead of having a separate (duplicate) coverage job? That will reduce errors (like the integration coverage test not actually running the integration tests), plus collecting coverage for each version will catch cases where we may have python-version-specific code.

@jsiirola I think these are all very good points. For more context, I can elaborate briefly on the reasons that led me to the current setup:

- When possible, I tried to follow the CircleCI configuration, where the coverage step was run outside of the test matrix (i.e., for a single version of Python)
- Performance: running tests with coverage adds a significant overhead (eyeballed from the test runtimes, up to 80% more) compared to the same tests without coverage
  - This might not be a critical point: since the jobs run in parallel, the overall runtime is determined by the slowest job (which, because of the overhead, will always be the coverage job)
  - However, it might result in longer overall runtimes if e.g. the coverage is run for test matrix configurations that already take longer without coverage: e.g. I've seen anecdotally that tests on Windows runners take longer than the same tests on Linux (even though the difference seems less pronounced with latest runs)
- This is a more conceptual point, but I can see an argument for considering the test coverage report being slightly different than just running the unit tests
  - i.e., instead of being a runtime check (like running the test suite by itself), it's more of an overall observable of the codebase (in this regard, more similar to the result of static analysis tools such as e.g. pylint)
  - In this sense, it functions more as a separate check that takes a long time to run which is independent from and complementary to the actual integration tests: this is what motivated the choice of having the coverage as a separate job
- Lastly, related to this, is the reason why the coverage is calculated on the complete test suite instead of only on the integration tests: my thinking was that, since the `-m integration` tests and the `-m 'not integration'` tests are disjoint sets, the complete test suite will have the largest possible coverage, which is ultimately the measure we'd want to know at this stage of a PR

**Summary**: This is what made sense to me while working on this, but I don't have much "real life" experience on this and I could be missing important aspects. Therefore I don't have strong opinions on preferences on the current configuration or the more general arguments and I'm happy to make any of the changes that you're suggesting.

File: .github/workflows/pr-approved.yml
Comment (lbianchi-lbl) [reply .github/workflows/pr-approved.yml:68]: See comment for more context: https://github.com/IDAES/idaes-pse/pull/110#issuecomment-743371850

**Comment by jsiirola** [comment]
One note:
> - Lastly, related to this, is the reason why the coverage is calculated on the complete test suite instead of only on the integration tests: my thinking was that, since the -m integration tests and the -m 'not integration' tests are disjoint sets, the complete test suite will have the largest possible coverage, which is ultimately the measure we'd want to know at this stage of a PR

One of the nice things about codecov is that it collects multiple coverage reports and shows / reports on the union of the reports.  So, since the `'not integration'` tests are initially reported by the PR, then the approval job only needs to "fill in the blanks" with the coverage from the `'integration'` tests.

**Review by jsiirola — APPROVED** [review]

**Comment by lbianchi-lbl** [comment]
> One note:
> 
> > * Lastly, related to this, is the reason why the coverage is calculated on the complete test suite instead of only on the integration tests: my thinking was that, since the -m integration tests and the -m 'not integration' tests are disjoint sets, the complete test suite will have the largest possible coverage, which is ultimately the measure we'd want to know at this stage of a PR
> 
> One of the nice things about codecov is that it collects multiple coverage reports and shows / reports on the union of the reports. So, since the `'not integration'` tests are initially reported by the PR, then the approval job only needs to "fill in the blanks" with the coverage from the `'integration'` tests.

Ah, I see. This is an excellent point and something I haven't taken into account while thinking about this. I think that now I fully understand your comments about adding the coverage report to all tests in the OS/Python version matrix:

- The final coverage report will contain the merged set of the lines from each test run; therefore, running the coverage report for all the versions we're testing will give us the most complete picture of what lines are covered, as it will include also the portions of the code that are specific to a particular OS or Python version
- If (as I understand) Codecov creates a report and then adds to it any new information reported by successive partial runs, then there's no need to run the complete test suite separately to have the complete test report: the coverage results from different partitions of the test suite can be run at different times, but they will all contribute to a single common report containing the combined results from all the partial runs

If my understanding is correct, then I agree with your initial suggestion of simply adding the coverage flag and report generation directly to the jobs running pytest for the test matrix:

- The performance/runtime argument does not apply, since the test run in parallel and the workflow will only be as fast as its slowest job
- For the integration tests, the overall runtime should end up decreasing, since we won't need to run the separate coverage run on the complete test suite (which is likely the job with the longest runtime, assuming the differences between Windows and Linux are not large)
- The upside is that the coverage report will be more complete, since it will include all of the code and all of the tests covering it
  - The difference will depend on the fraction of platform/version-specific code in the codebase, which I'm not yet familiar with; if I had to guess, I'd suspect it to be not dramatic, but not insignificant either

Let me know if you spot any misunderstandings in this summary. Otherwise, I think it makes sense to apply the changes you suggested because of the improved coverage and the workflows being a bit simpler.

Thanks a lot for all the review comments, BTW - I'm sorry it's been going back and forth for a while, but this information has been very valuable, both for me to learn, and for the CI workflows to improve.

**Review by dangunter — COMMENTED** [review]
LGTM

**Review by dangunter — APPROVED** [review]
LGTM

**Pull request merged**
